### ISSUE TITLE
Feature request: Native support for float16_t and related types

### ISSUE THREAD

**Issue by smileyofoz** [danielscherzer/GLSL#102]
### Installed product versions
- Visual Studio: 2022 Community
- This extension: GLSL Language Integration version 0.11.158

### Description
This is a feature request for native support for intellisence for the float16_t (and related types).

### Steps to recreate
1. Replace this
2. text with 
3. the steps
4. to recreate

### Current behavior
I'm using the glsl extension GL_AMD_gpu_shader_half_float.  Although the shader happily compiles to spir-v using glslangValidator, I get squiggly lines and VS2022 intellisense complains about the type with a syntax error.  The shader program also executes flawlessly.

### Expected behavior
It would be wonderful to have the float16_t (and f16vec2 et. al.) supported natively.

**Comment by danielscherzer** [comment]
Thanks for your feedback! Could you provide me with a short example shader that produces these errors?

**Comment by smileyofoz** [comment]
Hello Daniel!

Thanks so much for your reply.  May I first start by saying how much I love
this tool and how it changed my shader development experience overnight!

A code fragment that causes this behaviour is:

#version 460 core

#extension GL_AMD_gpu_shader_half_float : enable
#extension GL_AMD_gpu_shader_int16 : enable

layout (location = 0) out vec4 colour;

void main()
{
//Some meaningless declarations but it will reproduce the issue
        //Leave this next line uncommented to see it on uint16_t
uint16_t ui16; //Comment this line to see the squiggles on float16_t instead
float16_t f16;

vec4 colour = vec4(1., 0., 0., 1.);
}

I have some screenshots of the behaviour also.  You'll notice that as I
comment out the first uint16_t variable declaration, the squiggles show up
on the float16_t.  Also, it may be that a file that is included and makes
use of the //? syntax causes the error to present differently (the image
with the "Characteristic" struct demonstrates this).

[image: float16_t.png]
[image: Included_uin16_t.png]
[image: uint16_t.png]

Again, my shaders compile to spir-v just fine.
I'm using:
- version 0.11.158 of the extension
- VS2022 version 17.1.5
- Nvidia RTX-2080 Super
- Nvidia driver version 512.15

Thank you so much for your help, Shane

**Comment by danielscherzer** [comment]
Thanks for the praise!

Ah ok  I see the problem and the solution:
As a default the extension tries to compile your shader with the glsl compiler of the OpenGL driver of your graphics card. 
For instance my graphic driver (and it seems yours too) (NV2080) does not support the uint16_t for OpenGL.
Solution: Use an external compiler like glslangValidator. 
Info on how to: https://github.com/danielscherzer/GLSL#external-compiler

Short version: set the path to glslangvalidator in the extension options and specify arguments.
For my setup:
Arguments: -v
External compiler: D:\Tools\_Programming\glslang\bin\glslangValidator.exe

Then the squiggles disappear for me.
Does this help?

**Comment by smileyofoz** [comment]
Hello Daniel, thanks again for the reply mate.  I've tried setting the
external compiler up and I'm not convinced that it's being invoked:
[image: image.png]
The squiggles are still there.  If I try to misname the glslangValidator
executable to see if it complains (to see if it's being invoked) it seems
to fall back to the driver compiler, if it's trying to invoke it at all.

I normally invoke the compiler using this command:
glslangValidator %(FullPath) -G -o
$(TargetDir)Shaders\%(Filename)%(Extension).spv

Do I have this set up correctly?

Cheers, Shane

**Comment by smileyofoz** [comment]
So as I continue to dig, it seems that the squiggles (or any real-time
feedback) disappear altogether now because the external compiler is being
aborted.  Sorry to keep bugging you - just trying to arm you with useful
information :)

Cheers, Shane

**Comment by smileyofoz** [comment]
If I ctrl-B the file to build it individually, the executable does its job
pretty expeditiously.  It just seems to have issues when it's doing it
"intellisense" style.  Cheers

**Comment by danielscherzer** [comment]
Could you try:
Argument: -v
External compiler: full path to external compiler.

**Comment by smileyofoz** [comment]
Hi Daniel, I did so and the result I get is that the tool never seems to
end, but it does not complain about the 10s timeout and abort, either:
[image: image.png]

PS I applied the update that was released today, if that is of interest :)

Cheers and thank you so much for continuing to work with me!

**Comment by smileyofoz** [comment]
Hello again Daniel!

I'd like to make a feature request please.  When typing comments, the
keyword completion remains active, causing the plugin to insert keywords
from its list when I hit enter, instead of just entering a new line.  I
then need to go back and delete the inserted keyword. Would it be possible
to switch off auto-suggest when in comments or comment blocks?

Thank you, Shane

**Comment by smileyofoz** [comment]
Yes thank you Daniel - that would fulfill my needs perfectly! Cheers and
thanks again for such a great tool!

Shane

**Comment by smileyofoz** [comment]
Actually I see this in the bottom left hand side of the IDE:
[image: image.png]
I imagine this is related...  Cheers
